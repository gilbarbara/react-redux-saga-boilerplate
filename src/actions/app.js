// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
