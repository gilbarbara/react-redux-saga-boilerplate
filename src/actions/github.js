// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { githubGetRepos: getRepos } = createActions({
  [ActionTypes.GITHUB_GET_REPOS]: (query: string) => ({ query }),
});
