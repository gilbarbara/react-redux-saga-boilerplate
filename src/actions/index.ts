export * from './app';
export * from './github';
export * from './user';
