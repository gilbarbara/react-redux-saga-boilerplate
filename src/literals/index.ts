import { keyMirror } from '@gilbarbara/helpers';

import { Status } from '~/types';

export const ActionTypes = keyMirror({
  ALERTS_HIDE: undefined,
  ALERTS_SHOW: undefined,
  GITHUB_GET_REPOS_REQUEST: undefined,
  GITHUB_GET_REPOS_SUCCESS: undefined,
  GITHUB_GET_REPOS_FAILURE: undefined,
  SET_APP_OPTIONS: undefined,
  USER_LOGIN_REQUEST: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT_REQUEST: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined,
});

export const STATUS: Status = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
