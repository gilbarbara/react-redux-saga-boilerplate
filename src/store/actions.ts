export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
