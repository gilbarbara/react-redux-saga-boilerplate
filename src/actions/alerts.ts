import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from '~/modules/helpers';

import { ActionTypes } from '~/literals';

import { ShowAlertOptions } from '~/types';

export const hideAlert = createAction<string>(ActionTypes.ALERTS_HIDE);

export const showAlert = createAction(
  ActionTypes.ALERTS_SHOW,
  (message: ReactNode, options: ShowAlertOptions) => {
    const timeout = options.type === 'error' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'info-o',
      message,
      position: options.position || 'bottom-right',
      type: options.type || 'neutral',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
