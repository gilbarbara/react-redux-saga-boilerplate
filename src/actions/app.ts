/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
