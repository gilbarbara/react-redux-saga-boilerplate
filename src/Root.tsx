import { useEffect } from 'react';
import { Helmet } from 'react-helmet-async';
import { useDispatch } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { ThemeProvider } from '@emotion/react';
import styled from '@emotion/styled';
import { px } from '@gilbarbara/components';
import useTreeChanges from 'tree-changes-hook';

import { name } from '~/config';

import { useAppSelector } from '~/modules/hooks';
import theme, { headerHeight } from '~/modules/theme';

import { showAlert } from '~/actions';

import Footer from '~/components/Footer';
import Header from '~/components/Header';
import PrivateRoute from '~/components/PrivateRoute';
import PublicRoute from '~/components/PublicRoute';
import SystemAlerts from '~/containers/SystemAlerts';
import Home from '~/routes/Home';
import NotFound from '~/routes/NotFound';
import Private from '~/routes/Private';

import { selectUser } from '~/selectors';
import { UserState } from '~/types';

const AppWrapper = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  opacity: 1 !important;
  position: relative;
  transition: opacity 0.5s;
`;

const Main = styled.main<Pick<UserState, 'isAuthenticated'>>`
  min-height: 100vh;
  padding: ${({ isAuthenticated }) => (isAuthenticated ? `${px(headerHeight)} 0 0` : 0)};
`;

function Root() {
  const dispatch = useDispatch();
  const user = useAppSelector(selectUser);
  const { changed } = useTreeChanges(user);

  const { isAuthenticated } = user;

  useEffect(() => {
    if (changed('isAuthenticated', true)) {
      dispatch(showAlert('Hello! And welcome!', { type: 'success', icon: 'bell', timeout: 10 }));
    }
  }, [dispatch, changed]);

  return (
    <BrowserRouter>
      <ThemeProvider theme={theme}>
        <AppWrapper data-component-name="app">
          <Helmet
            defaultTitle={name}
            defer={false}
            encodeSpecialCharacters
            htmlAttributes={{ lang: 'pt-br' }}
            titleAttributes={{ itemprop: 'name', lang: 'pt-br' }}
            titleTemplate={`%s | ${name}`}
          >
            <link
              href="https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,400;0,700;1,400;1,700&display=swap"
              rel="stylesheet"
            />
          </Helmet>
          {isAuthenticated && <Header />}
          <Main isAuthenticated={isAuthenticated}>
            <Routes>
              <Route
                element={
                  <PublicRoute isAuthenticated={isAuthenticated} to="/private">
                    <Home />
                  </PublicRoute>
                }
                path="/"
              />
              <Route
                element={
                  <PrivateRoute isAuthenticated={isAuthenticated} to="/">
                    <Private />
                  </PrivateRoute>
                }
                path="/private"
              />
              <Route element={<NotFound />} path="*" />
            </Routes>
          </Main>
          <Footer />
          <SystemAlerts />
        </AppWrapper>
      </ThemeProvider>
    </BrowserRouter>
  );
}

export default Root;
