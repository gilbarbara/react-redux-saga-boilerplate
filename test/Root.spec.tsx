import React from 'react';
import { act, render, screen } from 'test-utils';

import { loginSuccess } from '~/actions';
import { ActionTypes } from '~/literals';

import Root from '~/Root';

const mockDispatch = vi.fn();

describe('Root', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should render properly for anonymous users', () => {
    render(<Root />, { mockDispatch });
    expect(screen.getByTestId('app')).toMatchSnapshot();
  });

  it('should render properly for logged users', () => {
    const { store } = render(<Root />, {
      mockDispatch,
    });

    act(() => {
      store.dispatch(loginSuccess());
    });

    expect(screen.getByTestId('app')).toMatchSnapshot();
    expect(mockDispatch).toHaveBeenCalledWith({
      type: ActionTypes.ALERTS_SHOW,
      payload: {
        id: '8cdee72f-28d4-4441-91f0-c61f6e3d9684',
        icon: 'bell',
        message: 'Hello! And welcome!',
        position: 'bottom-right',
        type: 'success',
        timeout: 10,
      },
    });
  });
});
