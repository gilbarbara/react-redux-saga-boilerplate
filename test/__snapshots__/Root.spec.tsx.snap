// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Root > should render properly for anonymous users 1`] = `
<div
  class="css-bv0go7"
  data-component-name="app"
>
  <div>
    Helmet
  </div>
  <main
    class="css-46qa52"
  >
    <div
      class="css-146gbh6"
      data-component-name="Home"
    >
      <div
        class="css-1ym6d5y"
        data-component-name="Container"
      >
        <div
          class="css-167s73c"
          data-component-name="Box"
        >
          <div
            class="css-1llpsq5"
          >
            <div
              class="css-11mttn6"
            >
              <svg
                src="/media/brand/icon.svg"
              />
            </div>
          </div>
          <h1
            class="css-2kvgnr"
          >
            React Redux Saga Boilerplate
          </h1>
          <button
            class="css-3qd3v6"
            data-component-name="Start"
            type="button"
          >
            <span
              class="css-1k4akb"
              data-component-name="Icon"
              name="star"
            >
              <svg
                height="24"
                src="https://files.gilbarbara.dev/icons/star.svg"
                title="star"
                width="24"
              />
            </span>
            <span
              class="css-iz9y3l"
              data-component-name="Text"
            >
              Start
            </span>
          </button>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="css-mtlhd8"
    data-component-name="Box"
  >
    <div
      class="css-fd9se2"
      data-component-name="Container"
    >
      <div
        class="css-1xy02bk"
        data-component-name="Box"
      >
        <iframe
          frameborder="0"
          height="20px"
          scrolling="0"
          src="https://ghbtns.com/github-btn.html?user=gilbarbara&repo=react-redux-saga-boilerplate&type=star&count=true"
          title="GitHub Stars"
          width="110px"
        />
        <iframe
          frameborder="0"
          height="20px"
          scrolling="0"
          src="https://ghbtns.com/github-btn.html?user=gilbarbara&type=follow&count=true"
          title="GitHub Follow"
          width="135px"
        />
      </div>
    </div>
  </footer>
  <div
    class="css-1ewh9kd"
  >
    <div
      class="css-vg32sl"
    >
      <div />
    </div>
    <div
      class="css-1r5pw8k"
    >
      <div />
    </div>
    <div
      class="css-50gqfe"
    >
      <div />
    </div>
    <div
      class="css-ypj8vw"
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`Root > should render properly for logged users 1`] = `
<div
  class="css-bv0go7"
  data-component-name="app"
>
  <div>
    Helmet
  </div>
  <header
    class="css-157to9t"
    data-component-name="Header"
  >
    <div
      class="css-1ejvhcb"
      data-component-name="Container"
    >
      <div
        class="css-11mttn6"
      >
        <svg
          src="/media/brand/icon.svg"
        />
      </div>
      <button
        class="css-17rt720"
        data-component-name="Logout"
      >
        <span
          class="css-lo0v0o"
          data-component-name="Text"
        >
          logout
        </span>
        <span
          class="css-10mt42n"
          data-component-name="Icon"
          name="sign-out"
        >
          <svg
            height="16"
            src="https://files.gilbarbara.dev/icons/sign-out.svg"
            title="sign-out"
            width="16"
          />
        </span>
      </button>
    </div>
  </header>
  <main
    class="css-gewp09"
  >
    <div
      data-to="/private"
    >
      NAVIGATE
    </div>
  </main>
  <footer
    class="css-mtlhd8"
    data-component-name="Box"
  >
    <div
      class="css-fd9se2"
      data-component-name="Container"
    >
      <div
        class="css-1xy02bk"
        data-component-name="Box"
      >
        <iframe
          frameborder="0"
          height="20px"
          scrolling="0"
          src="https://ghbtns.com/github-btn.html?user=gilbarbara&repo=react-redux-saga-boilerplate&type=star&count=true"
          title="GitHub Stars"
          width="110px"
        />
        <iframe
          frameborder="0"
          height="20px"
          scrolling="0"
          src="https://ghbtns.com/github-btn.html?user=gilbarbara&type=follow&count=true"
          title="GitHub Follow"
          width="135px"
        />
      </div>
    </div>
  </footer>
  <div
    class="css-1ewh9kd"
  >
    <div
      class="css-vg32sl"
    >
      <div />
    </div>
    <div
      class="css-1r5pw8k"
    >
      <div />
    </div>
    <div
      class="css-50gqfe"
    >
      <div />
    </div>
    <div
      class="css-ypj8vw"
    >
      <div>
        <div
          class="css-1cmax56 slide-left-enter slide-left-enter-active"
        >
          <div
            class="css-d735a"
            data-component-name="Alert"
            role="alert"
          >
            <span
              class="css-nri15g"
              data-component-name="Icon"
              name="bell"
            >
              <svg
                height="20"
                src="https://files.gilbarbara.dev/icons/bell.svg"
                title="bell"
                width="20"
              />
            </span>
            <span
              class="css-lo0v0o"
              data-component-name="Text"
            >
              <div
                class="css-185xs5d"
                data-component-name="Box"
              >
                Hello! And welcome!
              </div>
              <button
                class="css-1sdkvsj"
                data-id="8cdee72f-28d4-4441-91f0-c61f6e3d9684"
                type="button"
              >
                <span
                  class="css-fjs4pi"
                  data-component-name="Icon"
                  name="times"
                >
                  <svg
                    height="10"
                    src="https://files.gilbarbara.dev/icons/times.svg"
                    title="times"
                    width="10"
                  />
                </span>
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
